pallets/keystore/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+		#[pallet::weight(T::WeightInfo::revoke_keys(T::MaxKeys::get() as u32))]
+		pub fn revoke_keys(
+			origin: OriginFor<T>,
+			keys: Vec<T::Hash>,
+			key_purpose: KeyPurpose,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			ensure!(keys.len() > 0, Error::<T>::NoKeys);
+			ensure!(
+				keys.len() <= T::MaxKeys::get() as usize,
+				Error::<T>::TooManyKeys
+			);
+
+			for key in keys {
+				Self::revoke_key(account_id.clone(), key, key_purpose.clone())?;
+			}
+
+			Ok(())
+		}
+
+		/// Set a new key deposit.
+		#[pallet::weight(T::WeightInfo::set_deposit())]
+		pub fn set_deposit(origin: OriginFor<T>, new_deposit: T::Balance) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			<KeyDeposit<T>>::set(new_deposit);
+
+			Self::deposit_event(Event::DepositSet(new_deposit));
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Add a key to the `Keys` and `LastKeyByPurpose` storages if the following checks pass:
+		///
+		/// - The account has enough funds to cover the `key_deposit`;
+		/// - A key with the same hash and purpose does not exist in the `Keys` storage;
+		///
+		/// The `key_deposit` is reserved upon success.
+		fn add_key(
+			account_id: T::AccountId,
+			add_key: AddKey<T::Hash>,
+			key_deposit: T::Balance,
+		) -> DispatchResult {
+			T::Currency::reserve(&account_id, key_deposit)?;
+
+			let key_id: KeyId<T::Hash> = (add_key.key.clone(), add_key.purpose.clone());
+
+			<Keys<T>>::try_mutate(
+				account_id.clone(),
+				key_id.clone(),
+				|key_opt| -> DispatchResult {
+					match key_opt {
+						Some(_) => Err(Error::<T>::KeyAlreadyExists.into()),
+						None => {
+							let _ = key_opt.insert(Key {
+								purpose: add_key.purpose.clone(),
+								key_type: add_key.key_type.clone(),
+								revoked_at: None,
+								deposit: key_deposit,
+							});
+
+							Ok(())
+						}
+					}
+				},
+			)?;
+
+			<LastKeyByPurpose<T>>::insert(
+				account_id.clone(),
+				add_key.purpose.clone(),
+				add_key.key.clone(),
+			);
+
+			Self::deposit_event(Event::KeyAdded(
+				account_id.clone(),
+				add_key.key.clone(),
+				add_key.purpose.clone(),
+				add_key.key_type.clone(),
+			));
+
+			Ok(())
+		}
+
+		/// Revoke a key at the current `block_number` in the `Keys` storage
+		/// if the key is found and it's *not* already revoked.
+		fn revoke_key(
+			account_id: T::AccountId,
+			key: T::Hash,
+			key_purpose: KeyPurpose,
+		) -> DispatchResult {
+			let key_id: KeyId<T::Hash> = (key, key_purpose);
+
+			<Keys<T>>::try_mutate(
+				account_id.clone(),
+				key_id,
+				|storage_key_opt| -> DispatchResult {
+					match storage_key_opt {
+						Some(storage_key) => {
+							if let Some(_) = storage_key.revoked_at {
+								return Err(Error::<T>::KeyAlreadyRevoked.into());
+							}
+
+							let block_number = <frame_system::Pallet<T>>::block_number();
+
+							storage_key.revoked_at = Some(block_number.clone());
+
+							Self::deposit_event(Event::KeyRevoked(
```

Comment:
`nit`

_personally_ I prefer to keep lines of code that are part of the same action together to reduce visual fragmentation but this is fine of course.

```suggestion
							let block_number = <frame_system::Pallet<T>>::block_number();
							storage_key.revoked_at = Some(block_number.clone());
							Self::deposit_event(Event::KeyRevoked(
```